pkg/integrations/manager.go

```diff
@@ -116,155 +113,238 @@ func (c *ManagerConfig) DefaultRelabelConfigs() ([]*relabel.Config, error) {
-	integrations := make(map[Config]Integration)
-
-	for _, integrationCfg := range c.Integrations {
-		if integrationCfg.CommonConfig().Enabled {
-			l := log.With(logger, "integration", integrationCfg.Name())
-			i, err := integrationCfg.NewIntegration(l)
-			if err != nil {
-				return nil, err
-			}
-			integrations[integrationCfg] = i
-		}
-	}
+	ctx, cancel := context.WithCancel(context.Background())
 
-	return newManager(c, logger, im, integrations, validate)
-}
+	m := &Manager{
+		logger: logger,
+
+		ctx:    ctx,
+		cancel: cancel,
+
+		im:        im,
+		validator: validate,
+
+		integrations: make(map[string]*integrationProcess, len(c.Integrations)),
+	}
 
-func newManager(c ManagerConfig, logger log.Logger, im instance.Manager, integrations map[Config]Integration, validate configstore.Validator) (*Manager, error) {
-	defaultRelabels, err := c.DefaultRelabelConfigs()
+	var err error
+	m.hostname, err = instance.Hostname()
 	if err != nil {
-		return nil, fmt.Errorf("cannot get default relabel configs: %w", err)
+		return nil, err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
+	if err := m.ApplyConfig(c); err != nil {
+		return nil, fmt.Errorf("failed applying config: %w", err)
+	}
+	return m, nil
+}
 
-	m := &Manager{
-		c:                     c,
-		logger:                logger,
-		integrations:          integrations,
-		defaultRelabelConfigs: defaultRelabels,
-		im:                    im,
-		cancel:                cancel,
-		done:                  make(chan bool),
-		validator:             validate,
+// ApplyConfig updates the configuration of the integrations subsystem.
+func (m *Manager) ApplyConfig(cfg ManagerConfig) error {
+	m.cfgMut.Lock()
+	defer m.cfgMut.Unlock()
+
+	m.integrationsMut.Lock()
+	defer m.integrationsMut.Unlock()
+
+	if cmp.Equal(m.cfg, cfg) {
+		return nil
+	}
+
+	select {
+	case <-m.ctx.Done():
+		return fmt.Errorf("Manager already stopped")
+	default:
+		// No-op
 	}
 
-	if c.UseHostnameLabel {
-		var err error
-		m.hostname, err = instance.Hostname()
+	// Iterate over our integrations. New or changed integrations will be
+	// started, with their existing counterparts being shut down.
+	for _, ic := range cfg.Integrations {
+		// Key is used to identify the instance of this integration within the
+		// instance manager and within our set of running integrations.
+		key := integrationKey(ic.Name())
+
+		// Look for an existing integration with the same key. If it exists and
+		// is unchanged, we have nothing to do. Otherwise, we're going to recreate
+		// it with the new settings, so we'll need to stop it.
+		if p, exist := m.integrations[key]; exist {
+			if cmp.Equal(p.cfg, ic) {
+				continue
+			}
+			p.stop()
+			delete(m.integrations, key)
+		}
+
+		l := log.With(m.logger, "integration", ic.Name())
+		i, err := ic.NewIntegration(l)
 		if err != nil {
-			return nil, err
+			return fmt.Errorf("error initializing integration %q: %w", ic.Name(), err)
```

Comment:
What happens when this returns an error? Is it a problem that new configuration has been applied to some integrations an not others? Or is the idea that the agent would exit if this happened?

Reply:
I've been thinking about this as I've been implementing these, and I see what you're getting at. It's a problem that this returns an early error, since it leaves the Agent in a hard to predict state, where all valid changes were applied, but all (potentially valid) pending changes were aborted after coming across an invalid change. I can change this to log each error and return the first error after iterating through all of them. 

An error returned here would eventually turn into a 4xx/5xx HTTP response for the `/-/reload` caller. 

Given we're going to call /-/reload a beta feature, I think we should commit to saying it always updates the state of components that received a valid configuration, but any invalid configuration will leave that component in an undefined 
state (either running with the last known config or shut down and no longer running). 

It's hard to commit to anything more than this, especially since it's not always possible to bring up the component with the new settings before shutting down the old one. One example of this is the Tempo receivers, which open up TCP/UDP ports and need to be shut down so the new config can take its place. 

I hacked up some quick docs below, let me know what you think with the (beta feature) commitment for what we can guarantee /-/reload can/can't do. 

> `/-/reload` is an endpoint that updates the various components of the Agent to reflect the newest config on disk. It is currently in beta, so please open issues for any problems you see or ways to improve the functionality, especially around 
> handing runtime errors. 
> 
> When `/-/reload` is invoked, the config file provided to the Agent's `-config.file` flag will be re-read. Invalid YAML, or 
> YAML that doesn't pass basic validation will be immediately rejected. Otherwise, all components will be updated to 
> match the state of the new config file. 
> 
> When updating component state, runtime errors can happen that wouldn't be caught when reading in the YAML. 
> For example, setting the WAL to an invalid directory would be caught at runtime, not when the config file is read. 
> Runtime errors that prevent a component from being update will leave that component in an undefined state, where it is 
> either completely shut down or running with its last known config. Runtime errors must be fixed by addressing the 
> problem in the configuration file and invoking `/-/reload` again. 
> 
> Runtime errors encountered while updating the state do not stop the update from completing, and all valid changes will
> always be applied. The Agent has many "components" that receive updates:
>
> 1. The Server config 
> 2. Each Prometheus instance 
> 3. Each Loki instance 
> 4. Each Tempo instance 
> 5. Each Integration
> 6. The scraping service (which is the clustering mechanism + the key value store)
>
> If a call to /-/reload returns an error, always examine the logs to learn exactly what went wrong and which components 
> have been shut down as a result of the runtime error.

Reply:
I think that's a good compromise. "the component with invalid configuration is in an undefined state but other things are fine" feels like the right tradeoff when a user provides a bad config. As long as users can fix the config and call `/-/reload` again this seems like very straightforward behavior.

Reply:
Is it possible/wise to reject the update entirely and re-apply the old config? IE should we treat this has a transactional update. If server config is in an invalid state, can that propagate into all the other components being in an undefined state, even if their individual config is correct?

Reply:
It's possible, and I can see arguments on both sides here. I've been using Prometheus as the reference implementation which does the same thing as what we're doing now. Having transactional reloads will be an implementation detail of /-/reload itself though, where it just re-applies the last valid config before returning the error back to the user. 

I have some concerns about transactional reloads being noisy at the cluster level if that's being used, but this is something that can be discussed later. 

> If server config is in an invalid state, can that propagate into all the other components being in an undefined state, even if their individual config is correct?

Kind of. It's hard to do this in a strict way, but other components depend on the listen address / port being valid and open when the server starts listening. So if the server config is invalid, stuff will just subtly break at runtime.

Reply:
Sounds reasonable, especially if we let it chill in Beta for a bit and get feedback, with the knowledge to users that the semantics could change.